### ISSUE TITLE
Async request sending mis-handles partial writes

### ISSUE THREAD

**Issue by psychon** [not-a-seagull/breadx#19]
The following async program does not work (added as `examples/my_test.rs` to commit f7945528de4c8a1d4abb365945d0e5c67da7bbcf:)
```rust
use breadx::{DisplayConnection, ImageFormat, Image, drawable };
use smol::Executor;

// async entry point
async fn entry() -> breadx::Result<()> {
    const WIDTH: usize = 997;
    const HEIGHT: usize = 1009;
    const MEMORY: [u8; WIDTH * HEIGHT * 4] = [0; WIDTH * HEIGHT * 4];

    let mut conn = DisplayConnection::create(None, None)?;
    let image = Image::new(
        &conn,
        None,
        24,
        ImageFormat::ZPixmap,
        0,
        &MEMORY[..],
        WIDTH,
        HEIGHT,
        32,
        None,
    ).unwrap();
    let root = conn.default_root();
    let gc = conn.create_gc(root, Default::default())?;
    let pixmap = drawable::create_pixmap(&mut conn, root, WIDTH as _, HEIGHT as _, 24)?;
    drawable::put_image_async(
        &mut conn,
        pixmap,
        gc,
        &image,
        0,
        0,
        0,
        0,
        WIDTH as _,
        HEIGHT as _,
    ).await?;
    Ok(())
}

fn main() -> breadx::Result<()> {
    // spawn the executor
    let ex = Executor::new();
    async_io::block_on(ex.run(entry())).unwrap();
    Ok(())
}
```
xtrace says that too much data is sent for a request...?
```
000:<:0002: 16: Request(53): CreatePixmap depth=0x18 pid=0x02c00002 drawable=0x00000761 width=997 height=1009
000:<: Warning: buffer filled!
000:<:0003:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=0 left-pad=0x00 depth=0x18
000:<:0004:  0: Request(0): UNKNOWN   
000:<: Warning: buffer filled!
000:<:0005:  0: Request(0): UNKNOWN  
```
Using the sync version works fine:
```diff
--- examples/my_test_before.rs	2021-01-09 08:23:58.604168523 +0100
+++ examples/my_test.rs	2021-01-09 08:24:01.652113247 +0100
@@ -23,7 +23,7 @@ async fn entry() -> breadx::Result<()> {
     let root = conn.default_root();
     let gc = conn.create_gc(root, Default::default())?;
     let pixmap = drawable::create_pixmap(&mut conn, root, WIDTH as _, HEIGHT as _, 24)?;
-    drawable::put_image_async(
+    drawable::put_image(
         &mut conn,
         pixmap,
         gc,
@@ -34,7 +34,7 @@ async fn entry() -> breadx::Result<()> {
         0,
         WIDTH as _,
         HEIGHT as _,
-    ).await?;
+    )?;
     Ok(())
 }
 
```
```
000:<:0002: 16: Request(53): CreatePixmap depth=0x18 pid=0x02c00002 drawable=0x00000761 width=997 height=1009
000:<: Warning: buffer filled!
000:<:0003:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=0 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0004:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=65 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0005:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=130 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0006:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=195 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0007:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=260 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0008:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=325 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0009:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=390 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:000a:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=455 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:000b:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=520 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:000c:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=585 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:000d:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=650 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:000e:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=715 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:000f:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=780 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0010:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=845 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0011:259244: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=65 dst-x=0 dst-y=910 left-pad=0x00 depth=0x18
000:<: Warning: buffer filled!
000:<:0012:135616: Request(72): PutImage format=ZPixmap(0x02) drawable=0x02c00002 gc=0x02c00001 width=997 height=34 dst-x=0 dst-y=975 left-pad=0x00 depth=0x18
```
<details><summary>Unrelated, but I wanted to mention it anyway</summary>

```
$ cargo build --all-targets --features=async                                                                                                               master
   Compiling breadx v0.1.11 (/tmp/breadx)
warning: unused import: `XidType`
 --> examples/basic.rs:5:36
  |
5 |     WindowClass, WindowParameters, XidType,
  |                                    ^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: 1 warning emitted

error[E0599]: no method named `get_visual_configs_immediate` found for struct `breadx::Display<NameConnection>` in the current scope
  --> examples/glx_testing.rs:22:20
   |
22 |     let cfg = conn.get_visual_configs_immediate(conn.default_screen_index())?;
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method not found in `breadx::Display<NameConnection>`

error[E0599]: no method named `get_fb_configs_immediate` found for struct `breadx::Display<NameConnection>` in the current scope
  --> examples/glx_testing.rs:33:20
   |
33 |     let cfg = conn.get_fb_configs_immediate(conn.default_screen_index())?;
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^ method not found in `breadx::Display<NameConnection>`

warning: unreachable expression
  --> examples/glx_testing.rs:46:5
   |
41 | /     loop {
42 | |         let ev = conn.wait_for_event()?;
43 | |         println!("{:?}", ev);
44 | |     }
   | |_____- any code following this expression is unreachable
45 | 
46 |       Ok(())
   |       ^^^^^^ unreachable expression
   |
   = note: `#[warn(unreachable_code)]` on by default

error: aborting due to 2 previous errors; 1 warning emitted
```

</details>

**Comment by psychon** [comment]
Enabling log output says
```
WARN  breadx::display::connection] Called send_packet_async for blocking UnixStream
```
And indeed, switching to `DisplayConnection::create_async(None, None).await?` makes the example work...??

**Comment by psychon** [comment]
Uhm. How is the async request sending code in `src/display/connection/unix.rs` supposed to work?
```rust
    conn.write_with(|conn| {
        let connfd = conn.as_raw_fd();
        send_msg_packet(connfd, data, fds)
    })
    .await?;
```
This does not keep state between calls. When `send_msg_packet` sends part of the `data` and then encounters `WouldBlock`, the next call will again start from the beginning and try to send everything again.

**Comment by not-a-seagull** [comment]
This is largely an issue associated with an API decision I made early on that I've come to regret. Objects that implement the `Connection` trait are expected to implement methods for both sync and async transport. Aside from needing an ugly `async` suffix on every function, it also creates two failure states: calling the sync function on an async connection, and calling the async function on a sync connection. The former can be easily resolved by calling `block_on` for the async future, but the latter is a bit harder. I actually had a hack to set the connection into non-blocking mode that's still in the code tree, but I'm about 99% sure it doesn't work. (I actually had it removed in another branch and replaced with an error, but I haven't gotten around to merging it.)

To remedy this API problem, I'm going to create two separate traits: `Connection` and `AsyncConnection`. Sync methods will be used for `Display<T> where T: Connection`, and async methods will be used for `Display<T> where T: AsyncConnection`. This solves the confusing API problem. As for the interrupted packets, I'll implement a way to keep track of what's being written in it.

**Comment by not-a-seagull** [comment]
This issue should be fixed via #23
